### ISSUE TITLE
Don't reject HTTP/2 requests without body in low threads mode

### ISSUE THREAD

**Issue by sbordet** [eclipse/jetty.project#572]
After the work done in #558, the possibility to avoid to reject HTTP/2 requests without a body (those whose `HEADERS` frame has `endStream=true`) came up.

Because this kind of requests does not need to consume `DATA` frames and therefore never block in reads, they are somehow "lighter" to handle when in low threads mode.

However, such requests may block writing leading again to a scenario where the server hangs:
- the selector thread dispatches to HTTP/2's `onFillable()`
- the last thread invokes `onFillable`, reads the request and because it is in low threads mode it may dispatch it (rather than rejecting it), queuing it into the thread pool
- the `onFillable()` thread reads 0 and returns
- the request has now a free thread and gets executed, invoking the application
- the application writes, but the session send flow control window is zero, so it blocks
- in order to unblock writes we would need to read `WINDOW_UPDATE` frames from the client, but we are now out of threads: while the selector thread detects a read ready event, there are no threads to call `onFillable()` (because HTTP/2's fillable callback `isNonBlocking()=false`)

In the end, a lot of care needs to be taken, since for HTTP/2 we need more threads than HTTP/1 because we have to guarantee that both the NIO selection and the HTTP/2 fill _for each connection_ can always run. 

After #558, this is guaranteed by rejecting unconditionally every request when in low threads mode, but rejecting less requests may be really tricky.

**Comment by gregw** [comment]
I think it is important to move towards supporting this as I'm still cautious about the Http2Connection infillable task never being a non-blockable.    This means that with a mix of HTTP1 and HTTP2 connections, a low resources selector with always queue HTTP2 selections, while it will directly execute HTTP/1 unblocking actions.   HTTP2 connections can thus be stalled by HTTP1 actions.

I believe that HTTP2 onfillable should be non-blocking either always, or at least when low on resources.  Then when the server is low on resources,  window updates can always be handled, request without data frames always queued and only request with data frames rejected.

The challenge is that we want onfillable to be blocking if we are not low on resources so that we can get the EWYK mechanical sympathy.

Perhaps we need another class of dispatch which is optional non-blocking, so that when the selector strategy calls the task it can say then if it expects the task to be non-blocking or not?

**Comment by gregw** [comment]
The more I think about this, the more I am concerned that we are going to be rejecting requests too often.   With web pages often having 80+ nested resources, it will not be uncommon for a connection to dispatch a burst of 80 tasks, so it will only need a few simultaneous such connections to quickly exhaust the threadpool and we will start rejecting requests!  It is even conceivable that with a thread pool of 50 for an application that handles all it's request asynchronously, that we may still temporarily have 80+ outstanding jobs and thus reject some - even with a single connection!

The EWYK strategy was designed to avoid this kind of issue dispatching the producer rather than the consumer, so on a busy server, production would be slowed.   We will still get this behaviour because number of cores is much less than the number of threads (normally), but I think the current rejection strategy limits the goodness of this.

We should be able to dispatch jobs to a job queue that is deeper than 0 without too much concern..... but we are still stuck if multiple data frames in the dispatch queue consume all the 

So as well as trying to dispatch dataless requests, I'm still wondering if we should be looking sending a settings frame to limit the number of streams that can be sent - ie put back pressure on the client.
In reality, the HTTP2 protocol probably should have something in their flow control strategy for this.

Anyway, I'm currently thinking this is a higher priority that we thought otherwise to fix... and that it should be fixed in 9.3

**Comment by gregw** [comment]
I started writing this up as a formal http2 problem description... but soon devolved into jetty specific concerns.


The HTTP/2 flow control mechanism has been provided to limit an endpoint's resource commitment to a HTTP/2 connection. Specifically there is a per connection and per stream window that sets an upper limit of the number of bytes an endpoint will guarantee to read before stalling on the connection/stream.

The flow control window allows an end point to limit the total memory that must be dedicated to a connection/stream.   However there are other resources that must be allocated to a connection/stream which are limited and unable to be flow controlled. Specifically threads are required to service the messages received on a connection/stream.

The following is written from a servers point of view, but similar issues also exist client side.

Typically a HTTP/2 server will use a thread pool to limit the maximum number of threads that can be created in the server, as each thread consumes memory for stack and other OS resources.  However, there is limited ability for a server to apply back pressure based on the availability of threads and systems that support job queues in excess of the available threads are at risk of deadlock.

Consider the situation where a server is receiving many data streams over a single connection, each of them handled by an application that performs blocking reads on the stream.   It is possible for a dead lock to occur if all the available threads are blocked on reads that have been flow controlled due to an exhausted connection window, but that the data frames that have consumed that connection window have been received and placed in the job queue awaiting an available thread to consume the data.    A thread will never become available because they are all blocked on reads waiting for a data window that will never be opened because the data frames in the job queue will not be consumed until a thread is available.

So how can this situation be avoided by a server?

It can have an essentially infinite connection flow control window, ie a connection window that is larger than max-streams-per-connection * max-stream-window.   I believe this is what many browsers do to avoid this issue.  It works because any blocked stream is not dependent on progression of other stream, and thus deadlock cannot occur.

However, this approach essentially removes the ability to limit the memory commitment of a connection other than by limiting the number of streams that can be opened.  So is there something better that can be done?

In Jetty, we go to considerable lengths to ensure that if the server is low on threads, then the last thread available will continue to run select IO events from the connector and if any of those are non-blocking events, it will run them directly (eg unblocking blocked HTTP/1 reads and writes) and then only potentially blocking jobs are put on the job queued.    However, if there are HTTP/2 connections, we need to ensure that the last remaining non-blocked thread is used to run the connections so that h2 frames can be consumed.  Specifically, we need to ensure that we read and act on window-update frames so that any h2 write blocked streams can progress and free up their threads.    However, while reading frames looking for window updates, the last thread may also read header and data frames that have to either be handled, queued or rejected:

1. a data frame is received for an already read blocked stream, it should be handled immediately by being placed on the streams input queue and unblocking the read.

2. a data frame is received for a stream that has an asynchronous read pending, it cannot be handled as either the last thread must call the application callback (and potentially block) or the callback is put on the job queue, risking deadlock as now an allocation of connection flow control window is on the job queue.

3. A data frame is received for a stream that has not yet done a read (perhaps it has not yet been dispatched).  It is not safe to queue this frame on the HttpInput, as there is no guarantee that a thread will become available to consume the frame and thus flow control window may be held.   

4. A headers frame is received with no data.  This can safely be put on the job queue as it does not hold any allocation of flow control windows and thus can wait for an available thread without blocking any other stream from progressing.

5. A headers frame is received with expected data.   If this is put on the job queue, it is to be expected that if it gets a thread, it will likely block on a read.  But as the headers alone do not hold an allocation of the flow control window, this is probably OK but will eventually result in 2. or 3. above.

Jetty is currently avoiding the situation by resetting the stream when we hit 4. or 5. above.  With considerable effort, we could probably change this so that we allow 4. and 5. to proceed and only reset the stream if we hit 2. or 3. above.. 

Either way, resetting a stream is rather drastic and since a connection can easily have a burst of 80+ resources to fetch, it makes it highly likely that a page will not be able to completely load if a server has < 100 idle threads available.

We could look at sending settings frames to limit the number of streams, but I believe this would significantly limit the statistical resource sharing done by servers - essentially servers over-commit to many connections on the assumption that they will not all be busy at once.   If a server had to limit the total possible new stream that could arrive to the number of idle threads available, then total streams possible would be very low.

So currently I am thinking that the best way to proceed is to not reset, but instead disable the connection flow control window by forcing it to be > max-streams-per-connection * max-stream-window.  I believe that solves the read blocking issue.

I then think we need to revisit the write blocking issue, because we backed out the change making the http2 connection be able to be non-blocking, so the last thread in the selector could still be used to process a window update frame.      If the http2 connection is never non-blocking, then I believe we are vulnerable to a write deadlock again.

**Comment by gregw** [comment]
I think I can also argue that HTTP2 connections need not be run with EWYK/EPC scheduling.

The EWYK scheduling has excellent mechanical sympathy for HTTP/1, because the framing of the message required the detail of the message to be parsed - ie each header line needs to be parsed in detail to work out the framing of the message.  Thus once the framing is done, the message is parsed and the cache is hot with all the detail, so it is important to use the same thread/core to process.

But this is not the case with HTTP/2 framing, because it has a small well defined framing header around the contents of the header frame.     Thus once the frame has been parsed and identified as a headers frame, only a few bytes at the start of the frame will have been read into cache memory.   Thus the cost of dispatching to a new thread is much less, as that thread will be the first to read the contents of the header frame, thus the first to parse/cache the detail.

I think we should consider make Http2Connection always run in PEC mode, with the optimization that frames that are known not to be blocking can be executed by the producer - window updates, data frames queued on HttpInput etc.

**Comment by gregw** [comment]
hmmm hpack makes this analysis a little more complex as headers must be parsed in order....

**Comment by sbordet** [comment]
After dee3331, HTTP/2 runs by default with PEC and a non-blocking fill callback.
I also added setters to enlarge the server session recv window, but left it at 65535 by default.

**Comment by gregw** [comment]
Let's keep this issue open and discuss how we can make http2 in 9.4 avoid PEC, which has bad mechanical sympathy.   From internal email I suggested:

To avoid the race, I think it is better if we explicitly allocated the last thread(s), ie we take 1 or more threads from the thread pool and they are the ones that will act when the server is low on resources.    The rest of the thread pool can then be used with pure EWYK strategy and it will only be the low resource thread(s) that run the PEC strategy.

This is easy to achieve for a blocking producer like the selector manager, as you simply set a low priority on the low resource threads so that they only run if no other thread is running.  If they do run because a normal thread is blocked in produce, then they too will block (either on the lock or perhaps on a condition... need to think about that).   So that normally the server will run in EWYK, even if it has a long job queue.  Only if no threads can proceed (ie potentially deadlocking) will the low priority low resources thread run and do the selecting in PEC mode, which will unblock the normal threads.

I can definitely imagine how this can be made to work for selectors, as there are only few of them and 1 extra thread per selector is not a huge problem.

Making this work for http2 is more difficult, as there are many instances of execution strategy, one per connection, so we can't have a thread each.     Thus we need to come up with a better idea of how a low resources thread(s) can do the producing for many connections.... yet not spin if there are not tasks.

So there is more thinking to do... but essentially I think there is benefit in changing our thinking.  Rather than try to change strategies dynamically by guessing if we are low on threads, we should run both strategies and use thread priority to keep the preferred strategy as the prime actor.

**Comment by gregw** [comment]
OK I think I'm getting closer on this.    We need tasks that can be produced to declare if they are blocking, nonblocking or potentially either:

````java
public interface NonBlockingTask extends Runnable
{
    void nonBlockingRun();
    @Override
    default void run() { nonBlockingRun();}
}

public interface BlockingTask extends Runnable
{
    void blockingRun();
    @Override
    default void run() { blockingRun();}
}
````

So with both of those interfaces, it is possible for HttpConnection to declare that it can be run either blocking or non blocking mode (the caller can choose which or just call run if it doesn't care):
````java
public static class Http2ConnectionFillable implements NonBlockingTask, BlockingTask
{
    public void run() { blockingRun();}

    @Override
    public void blockingRun()
    {
        // call EPC.execute();
    }

    @Override
    public void nonBlockingRun()
    {
        // call PEC.execute
    }
}
````

Now the EPC implementation can have all the low resource and rejectable machinery removed from it.   It can then be setup to prefer calling Blocking tasks:

````java
private void executeProduceConsume()
{
    while (true)
    {
        Runnable task = _producer.produce();

        boolean dispatch = false;
        try (Lock locked = _locker.lock())
        {
            _producing = false;

            if (task == null)
            {
                if (_execute)
                {
                    _idle = false;
                    _producing = true;
                    _execute = false;
                    continue;
                }

                _idle = true;
                break;
            }

            // We have a task, which we will run ourselves which may block,
            // so if we don't have another thread pending
            if (!(task instanceof NonBlockingTask) && !_pending)
            {
                // dispatch one
                dispatch = _pending = true;
            }

            _execute = false;
        }

        if (dispatch)
        {
            if (!execute(this) && !(task instanceof NonBlockingTask))
                task = null;
        }

        // Run the task.
        if (task != null)
        {
            // Prefer blocking task if there is a choice.
            if (task instanceof BlockingTask)
                ((BlockingTask)task).blockingRun();
            else
                task.run();
        }

        // Once we have run the task, we can try producing again.
        try (Lock locked = _locker.lock())
        {
            // Is another thread already producing or we are now idle?
            if (_producing || _idle)
                break;
            _producing = true;
        }
    }
}
````

The PEC strategy is setup the otherway, to prefer NonBlocking tasks:

````java
// Produce until we no task is found.
while (true)
{
    // Produce a task.
    Runnable task = _producer.produce();

    if (task == null)
    {
        try (Lock locked = _locker.lock())
        {
            switch(_state)
            {
                case IDLE:
                    throw new IllegalStateException();
                case PRODUCE:
                    _state=State.IDLE;
                    return;
                case EXECUTE:
                    _state=State.PRODUCE;
                    continue;
            }
        }
    }

    // Run or Execute the task.
    if (task instanceof NonBlockingTask)
        ((NonBlockingTask)task).nonBlockingRun();
    else
        execute(task);
}        
````

If we really want, we can write both strategies to be configurable if they prefer Blocking or NonBlocking tasks.

So now Http2 onfillable can be registered as both a blocking and nonblocking callback (we need to reflect the same differences in Callbacks).     If the selectors EPC strategy is running, it will call the connections blockingRun() method, which will also use EPC to read, parse and handle the tasks.

If the server becomes busy and deadlocks in EPC vs job queue, then the low priority PEC thread will start working, it will select the Http2 connection, see it can be called non-blocking and call it directly, so the HTTP PEC run will be called to read parse and handle WINDOW UPDATE frames, all other frames are put on the job queue.

I'm pretty sure this is going to work..... not 100% happy with the names.

There is also the complication of what should the selector producer do if it has both a read and write callback and they are different combinations of blocking/nonblocking.  I think they need to be returned as separate tasks rather than a single task that calls both.

**Comment by gregw** [comment]
I've had a couple of attempts at implementing this.... I'm still convinced that it will work, but so far I've not like the API noise that it has introduced.

The current Callback API already has 
````java
boolean isBlockingCallback();
````
So it makes sense to change this to something like:
````java
enum ExecutionType { UNKNOWN, BLOCKING, NON_BLOCKING, EITHER };
ExecutionType getExecutionType();
````
So we can correctly ask a callback what kind of execution does it support (Note that "execution" may be the wrong word here as that may imply use of an executor.. open to suggestions? RunType? )

This works fine for the types `UNKNOWN`, `BLOCKING`, `NON_BLOCKING`, which are declarative in nature and we can just call succeeded() or failed() on the callback (or run() on the equivalent runnable) and we will get the declared behaviour.

The problem comes with type `EITHER`, which says the callback (or runnable) can be called in either blocking or non-blocking mode and it is the caller that decides which.

So one of my attempts was to put `succeededBlocking()` `succeededNonBlocking()` `failedBlocking(x)` `failedNonBlocking(x)` methods onto callback and `runBlocking()` `runNonBlocking()` methods onto Task extends Runnable.    I think this can be made to work, but it is a lot of API calls that have to be plumbed together, which gets verbose with all our callback wrappers and a bit of a brain bender when we need to convert from a callback to a runnable etc.     There are also lots of instanceof tests, coercions and the ability to call APIs in a way that needs an ISE or USOE to be thrown.

So I tried keeping a single method but passing in the ExecutionType.   It is a little better but still puts a lot of code in the way of the majority of simple executions.

Thus my current thinking is to use a ThreadLocal to anoint the calling thread as a NonBlocking caller, so only Runnables and Callbacks that implement Either need to consult the ThreadLocal to determine which mode to run in.     This has the advantage that all the existing plumbing throw wrapped callbacks etc. will work.      The cost will be a ThreadLocal set, lookup and then clear for every invocation of HTTP2 onfillable.   Is this an acceptable cost?

**Comment by gregw** [comment]
@sbordet what are your thoughts about a thread local vs a big API change?

**Comment by sbordet** [comment]
`ThreadLocal` works for me. I am interested in what kind of API change you have in mind.

**Comment by gregw** [comment]
@sbordet API changes were along the lines of my comment above.   Was a bit too verbose, so will try with thread local....

**Comment by gregw** [comment]
The PR #614 has implemented the low priority alternate strategy approach.   However the SmallThreadPoolLoadTest is still failing, but it does not immediately look like it is a pure thread starvation problem.  So that test has been ignored for now, but needs to be analysed before this issue is closed.

**Comment by gregw** [comment]
@sbordet where are we on this ?  we are good in 9.4 right?

**Comment by sbordet** [comment]
@gregw you have a much firmer grip on this than me.

I reviewed a few days ago class `Invocation`, and it really needs some Javadocs. I could not write them for `InvocationType.EITHER` because I did not know exactly what to write.

We would need a thorough review together IMHO. After the review, I could write the Javadocs for all class involved and then feed the documentation pipeline.

**Comment by WalkerWatch** [comment]
@sbordet Let me know if can be of any assistance.

**Comment by gregw** [comment]
@sbordet I improved the javadoc a little bit.   @WalkerWatch this is my description of the problem and our solution.

Jetty has several instances of a computing pattern called Produce-Consume, where a task is run that produces other tasks that need to be consumed.     An example of a producer is the HTTP/1.1 Connection, where the producer task looks for IO activity on a connection and any readable data detected is produced as a task, which is then Consumed by being handled by a Connection/Parser/Handler/Filter/Servlet etc.   In java NIO terms, the Producer in this example is running the NIO Selector an the Consumers are handling any IO that has been selected. 

How the Producer task and Consumer tasks are executed can have a dramatic effect on server performance due to mechanical sympathy issues.   For example, if one thread produces a task that is consumed by a different thread, then it is likely that a different CPU core may be used and the cache will be cold and poor execution will result.

Jetty has an ExecutionStrategy called ExecuteProduceConsume (nicknamed EatWhatYouKill) that has excellent mechanical sympathy properties (see https://webtide.com/eat-what-you-kill/), so we get very good performance.

But this raises another problem in jetty - thread starvation (when somebody else bites off more than they can chew :)    The issue is that ExecuteProduceConsume(EWYK)works by using the thread that is Producing to immediately consume any task that it Produces.   However, it is possible (even likely) that the task to be consumed may block (it's often calling application code), so the EWYK will dispatch another task to the thread pool that will do more producing.

The problem is that if the thread pool is empty (because all the threads are in blocking application code) then the last non-blocked thread may be Producing and produce a task which it then calls and it too Blocks.   The task to do more Produced has been dispatched the the thread pool, but is waiting in the job queue for an available Thread.    BUT all the threads are blocking and it may be that they are blocking on IO operations that will only be unblocked if some data is read/written.... but unless something calls the NIO selector, the read/write will not been seen.   Since the Selector is called by the Producer task and that is waiting in the queue and the queue is stalled because of all the threads blocked waiting for the selector - the server is now dead locked!

So, Jetty's clever solution is to not only run our EWYK execution strategy, but to also run the alternative (more traditional) ProduceExecuteConsume strategy, where 1 thread does all the producing and always dispatches any produced tasks to the thread pool.   Because this is not mechanically sympathetic, we run the producer task at low priority.    This effectively reserves one thread from the thread pool to always be a producer, but because it is low priority it will seldom run unless the server is idle - or completely stalled due to thread starvation!       This means that jetty always has a thread available to Produce, thus there is always a thread available to run the NIO Selector and any IO events that will unblock any threads will be detected.   This needs one more trick to work, the producing task must be able to tell if a detected IO task is non-blocking (ie a wakeup of a blocked read or write), in which case it executes it itself rather than submitting the task to any execution strategy.    Jetty uses the InvocationType interface to tag such tasks and thus avoid thread starvation.

This is a great solution when a thread can be dedicated to always Producing (eg NIO Selecting).  However Jetty has other Producer Consumer patterns that cannot be threadful.   HTTP2 Connections are consumers of IOEvents, but are themselves producers of parsed HTTP/2 frames which may be handled in parallel due to the multiplexed nature of HTTP/2.   So each HTTP/2 connection is itself a Produce-Consume pattern, but we cannot allocate a Producer thread to each connection as a server may have many tens of thousands connections!

Yet to avoid Thread starvation, we must also always call the Producer task for HTTP/2 as it may parse HTTP/2 flow control frames that are necessary to unblock the IO being done by applications threads that are blocked and holding all the available threads from the pool.

So even if there is a thread reserved as the Producer/Selector by a connector, it may detect IO on a HTTP/2 connection and use the last thread from the thread pool to Consume that IO, which if it the produces a HTTP/2 frame and EWYK strategy is used, then the last thread may Consume that frame and it too may block in application code.   So even if the reserved thread detects more IO, there are no more available threads to consume them!


So the solution in HTTP/2 is similar to in the Connector.   Each HTTP/2 connection has two executions strategies - EWYK which is used when the calling thread (the Connectors consumer) is allowed to block and the traditional ProduceExecuteConsume strategy, which is used when the calling thread is not allowed to block.    The HTTP/2 Connection then advertises itself as an InvocationType of EITHER to the Connector.    If the connector is running normally, it's EWYK strategy will be used and then the HTTP2 Connection can also use its efficient EWYK strategy.     However, if the Connector is running its low priority ProduceExecutionConsume strategy, it invokes the HTTP2Connection as non-blocking!   This tells the Http2 Connection that when it is acting as a Consumer of the Connectors task, it must not block - so it uses its own ProduceExecuteConsume strategy, as it new the Production will just parse the HTTP2 frame and not itself do the Consume task (which may block).

The final part is that the HTTP2 frame Producer can look at the frames produced, and if they are not frames that will block when handled (eg Flow Control) they are handled by the Producer and not submitted to any strategy to be Consumed.    Thus even if the Server is on it's last thread,  Flow Control frames will be detected, parsed and handled - thus unblocking other threads and avoiding starvation!


I think I should probably write this up in a blog???

**Comment by olamy** [comment]
+1 for blog post!

**Comment by WalkerWatch** [comment]
@gregw Fantastic write up and I think it would be an excellent blog post. If there is anything I can help with on that front, let me know.

**Comment by gregw** [comment]
A couple more Callbacks and runnables in SSL need to be tagged as EITHER or NON_BLOCKING

**Comment by sbordet** [comment]
I have review and edited the main comment, fixing a few errors and make it clearer (at least to me).

**Comment by gregw** [comment]
@sbordet you edited the comment in the issue or in the code?  I can't see any commit from you so guess in this issue?

**Comment by sbordet** [comment]
I edited the long comment you wrote above.
